include/boost/compute/program.hpp

```diff
@@ -516,7 +516,8 @@ class program
                 << "// " << options << "\n\n"
                 << source;
 
-            hash = detail::sha1(src.str());
+            // Gets rebuilt if either source, or device (indicated by name) or compile options change
+            hash = detail::sha1(src.str() + context.device.name().str() + options.str());
```

Comment:
As Denis commented, this additional information should already be stored inside the program source and therefore I don't think we need to store it here.